mithril-signer/src/services/epoch_service.rs

```diff
@@ -0,0 +1,335 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use mithril_common::entities::Epoch;
+use mithril_common::entities::ProtocolParameters;
+use mithril_common::entities::Signer;
+use mithril_persistence::store::StakeStorer;
+use slog_scope::{debug, trace};
+use thiserror::Error;
+
+use mithril_common::entities::EpochSettings;
+use mithril_common::entities::SignerWithStake;
+use mithril_common::StdResult;
+
+use crate::RunnerError;
+
+/// Errors dedicated to the CertifierService.
+#[derive(Debug, Error)]
+pub enum EpochServiceError {
+    /// Raised when service has not collected data at least once.
+    #[error("Epoch service was not initialized, the function `inform_epoch_settings` must be called first")]
+    NotYetInitialized,
+}
+
+/// Service that aggregates all data that don't change in a given epoch.
+#[async_trait]
+pub trait EpochService: Sync + Send {
+    /// Inform the service a new epoch has been detected, telling it to update its
+    /// internal state for the new epoch.
+    async fn inform_epoch_settings(&mut self, epoch_settings: EpochSettings) -> StdResult<()>;
+
+    /// Get the current epoch for which the data stored in this service are computed.
+    fn epoch_of_current_data(&self) -> StdResult<Epoch>;
+
+    /// Get next protocol parameters used in next epoch (associated with the actual epoch)
+    fn next_protocol_parameters(&self) -> StdResult<&ProtocolParameters>;
+
+    /// Get signers for the current epoch
+    fn current_signers(&self) -> StdResult<&Vec<Signer>>;
+
+    /// Get signers for the next epoch
+    fn next_signers(&self) -> StdResult<&Vec<Signer>>;
+
+    /// Get signers with stake for the current epoch
+    async fn current_signers_with_stake(&self) -> StdResult<Vec<SignerWithStake>>;
+
+    /// Get signers with stake for the next epoch
+    async fn next_signers_with_stake(&self) -> StdResult<Vec<SignerWithStake>>;
+}
+
+struct EpochData {
+    epoch: Epoch,
+    next_protocol_parameters: ProtocolParameters,
+    current_signers: Vec<Signer>,
+    next_signers: Vec<Signer>,
```

Comment:
Maybe the `current-signers_with_stake` and `next_signers_with_stake` should be stored in the `EpochData` to avoid recomputing them any time we need them (at least twice by signature round)?

Reply:
I'm not sure that we always can compute SignerWithStake when we create the EpochData because we need the stake distribution.
If it's the case, we need to implement a kind of lazy loading or a retry until data are available. 
We said that it was optimization that we could do in a second step.
Do you think we have do it in this PR ?